### ISSUE TITLE
LLVM 17.0.1 failed test on alpine linux: Transforms/SampleProfile/pseudo-probe-emit.ll

### ISSUE THREAD

**Issue by ncopa** [llvm/llvm-project#66776]
I'm trying to upgrade the llvm package for alpine linux but I get a regression test failure:

This is on alpine edge x86_64 (musl libc).

```
[1/1] Running the LLVM regression tests
FAIL: LLVM :: Transforms/SampleProfile/pseudo-probe-emit.ll (42301 of 48701)
******************** TEST 'LLVM :: Transforms/SampleProfile/pseudo-probe-emit.ll' FAILED ********************
Script:
--
: 'RUN: at line 2';   /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/bin/opt < /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/llvm/test/Transforms/SampleProfile/pseudo-probe-emit.ll -passes=pseudo-probe -function-sections -S -o /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/test/Transforms/SampleProfile/Output/pseudo-probe-emit.ll.tmp
: 'RUN: at line 3';   /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/bin/FileCheck /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/llvm/test/Transforms/SampleProfile/pseudo-probe-emit.ll < /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/test/Transforms/SampleProfile/Output/pseudo-probe-emit.ll.tmp --check-prefix=CHECK-IL
: 'RUN: at line 4';   /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/bin/llc /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/test/Transforms/SampleProfile/Output/pseudo-probe-emit.ll.tmp -stop-after=pseudo-probe-inserter -o - | /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/bin/FileCheck /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/llvm/test/Transforms/SampleProfile/pseudo-probe-emit.ll --check-prefix=CHECK-MIR
: 'RUN: at line 5';   /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/bin/llc /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/test/Transforms/SampleProfile/Output/pseudo-probe-emit.ll.tmp -function-sections -filetype=asm -o /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/test/Transforms/SampleProfile/Output/pseudo-probe-emit.ll.tmp1
: 'RUN: at line 6';   /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/bin/FileCheck /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/llvm/test/Transforms/SampleProfile/pseudo-probe-emit.ll < /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/test/Transforms/SampleProfile/Output/pseudo-probe-emit.ll.tmp1 --check-prefix=CHECK-ASM
: 'RUN: at line 7';   /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/bin/llc /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/test/Transforms/SampleProfile/Output/pseudo-probe-emit.ll.tmp -function-sections -filetype=obj -o /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/test/Transforms/SampleProfile/Output/pseudo-probe-emit.ll.tmp2
: 'RUN: at line 8';   /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/bin/llvm-readelf -S -g /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/test/Transforms/SampleProfile/Output/pseudo-probe-emit.ll.tmp2 | /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/bin/FileCheck /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/llvm/test/Transforms/SampleProfile/pseudo-probe-emit.ll --check-prefix=CHECK-SEC
: 'RUN: at line 9';   /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/bin/llvm-mc /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/test/Transforms/SampleProfile/Output/pseudo-probe-emit.ll.tmp1 -filetype=obj -o /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/test/Transforms/SampleProfile/Output/pseudo-probe-emit.ll.tmp3
: 'RUN: at line 10';   /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/bin/llvm-readelf -S -g /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/test/Transforms/SampleProfile/Output/pseudo-probe-emit.ll.tmp3 | /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/bin/FileCheck /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/llvm/test/Transforms/SampleProfile/pseudo-probe-emit.ll --check-prefix=CHECK-SEC
: 'RUN: at line 12';   /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/bin/llc /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/test/Transforms/SampleProfile/Output/pseudo-probe-emit.ll.tmp -function-sections -unique-section-names=0 -filetype=obj -o /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/test/Transforms/SampleProfile/Output/pseudo-probe-emit.ll.tmp4
: 'RUN: at line 13';   /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/bin/llvm-readelf -S /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/test/Transforms/SampleProfile/Output/pseudo-probe-emit.ll.tmp4 | /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/bin/FileCheck /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/llvm/test/Transforms/SampleProfile/pseudo-probe-emit.ll --check-prefix=CHECK-SEC2
--
Exit Code: 1

Command Output (stderr):
--
/home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/llvm/test/Transforms/SampleProfile/pseudo-probe-emit.ll:114:19: error: CHECK-SEC-NEXT: expected string not found in input
; CHECK-SEC-NEXT: .pseudo_probe PROGBITS {{.*}} 00 LG 8 0 1
                  ^
<stdin>:26:71: note: scanning from here
 [21] .pseudo_probe PROGBITS 0000000000000000 000146 000013 00 L 5 0 1
                                                                      ^
<stdin>:43:37: note: possible intended match here
COMDAT group section [ 12] `.group' [.pseudo_probe_desc_foo2] contains 1 sections:
                                    ^

Input file: <stdin>
Check file: /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/llvm/test/Transforms/SampleProfile/pseudo-probe-emit.ll

-dump-input=help explains the following input dump.

Input was:
<<<<<<
            .
            .
            .
           21:  [16] .note.GNU-stack PROGBITS 0000000000000000 0000c2 000000 00 0 0 1 
           22:  [17] .eh_frame X86_64_UNWIND 0000000000000000 0000c8 000058 00 A 0 0 8 
           23:  [18] .rela.eh_frame RELA 0000000000000000 000300 000048 18 I 22 17 8 
           24:  [19] .pseudo_probe PROGBITS 0000000000000000 000120 00000d 00 LG 8 0 1 
           25:  [20] .pseudo_probe PROGBITS 0000000000000000 00012d 000019 00 L 3 0 1 
           26:  [21] .pseudo_probe PROGBITS 0000000000000000 000146 000013 00 L 5 0 1 
next:114'0                                                                           X error: no match found
           27:  [22] .symtab SYMTAB 0000000000000000 000180 000138 18 1 9 8 
next:114'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           28: Key to Flags: 
next:114'0     ~~~~~~~~~~~~~~
           29:  W (write), A (alloc), X (execute), M (merge), S (strings), I (info), 
next:114'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           30:  L (link order), O (extra OS processing required), G (group), T (TLS), 
next:114'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           31:  C (compressed), x (unknown), o (OS specific), E (exclude), 
next:114'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            .
            .
            .
           38:  
next:114'0     ~
           39: COMDAT group section [ 10] `.group' [.pseudo_probe_desc_foo] contains 1 sections: 
next:114'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           40:  [Index] Name 
next:114'0     ~~~~~~~~~~~~~~
           41:  [ 11] .pseudo_probe_desc 
next:114'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
           42:  
next:114'0     ~
           43: COMDAT group section [ 12] `.group' [.pseudo_probe_desc_foo2] contains 1 sections: 
next:114'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
next:114'1                                         ?                                               possible intended match
           44:  [Index] Name 
next:114'0     ~~~~~~~~~~~~~~
           45:  [ 13] .pseudo_probe_desc 
next:114'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
           46:  
next:114'0     ~
           47: COMDAT group section [ 14] `.group' [.pseudo_probe_desc_foo3] contains 1 sections: 
next:114'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           48:  [Index] Name 
next:114'0     ~~~~~~~~~~~~~~
           49:  [ 15] .pseudo_probe_desc 
next:114'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
>>>>>>

--

********************
********************
Failed Tests (1):
  LLVM :: Transforms/SampleProfile/pseudo-probe-emit.ll


Testing Time: 413.35s
  Skipped          :    15
  Unsupported      :  1833
  Passed           : 52918
  Expectedly Failed:   151
  Failed           :     1
ninja: job failed: cd /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/test && /usr/bin/python3.11 /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/./bin/llvm-lit -sv /home/ncopa/aports/main/llvm17/src/llvm-project-17.0.1.src/build/test
ninja: subcommand failed
```

**Comment by Endilll** [comment]
I hope we can prevent such regressions by setting up a musl buildbot.

**Comment by ncopa** [comment]
Is there anything I can do to troubleshoot it? This is a blocker for the the upgrade to 17 on alpine linux, and it would be nice to have a fix, or even a workaround.

**Comment by Endilll** [comment]
@ncopa It would definitely help if you can find a commit that made this test fail for you. It might be even possible to revert it for upcoming 17.x release.

**Comment by ncopa** [comment]
I git bisected it:

62d861422317b3e94a97706f07e8f1abce9a8135 is the first bad commit

**Comment by Endilll** [comment]
CC @MaskRay

**Comment by Endilll** [comment]
CC @MaskRay as commit author

**Comment by ncopa** [comment]
If I revert 62d861422317b3e94a97706f07e8f1abce9a8135 the test passes.

**Comment by tru** [comment]
I think we need MaskRay's input here. If it's a harmless test failure, I don't think we should revert it in 17.0.1, but if it's pointing out an underlying issue we definitely should consider reverting.

**Comment by ncopa** [comment]
Yes, of course. Makes sense. I am willing to help test other fixes as well.

On a side note, I find it interesting that the test passes on aarch64.
Maybe I should test it on other architectures while waiting for @MaskRay

**Comment by ncopa** [comment]
It seems like this specific test only fails on x86_64. It passes on aarch64, armv7, armv6, ppc64le, s390x and x86 (32 bit).
https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/51171

**Comment by ncopa** [comment]
I have reverted 62d861422317b3e94a97706f07e8f1abce9a8135 in the alpine linux llvm17  build for now. Would be nice to know if we should keep it reverted or if we should disable the test before we do the 3.19 release now in November.

**Comment by ncopa** [comment]
> I hope we can prevent such regressions by setting up a musl buildbot.

How do we set up a musl buildbot?

**Comment by andypost** [comment]
@ncopa I bet it's https://llvm.org/docs/HowToAddABuilder.html

**Comment by MaskRay** [comment]
Sorry, I just saw this issue. pseudoprobe has non-determinism issues even before 62d861422317b3e94a97706f07e8f1abce9a8135, which I fixed in 001af0f89405a4397773970995a8d8d8328d4235 (2023-09) and another commit. Could the failure be related to non-determinism?

18.1 has been released. Have you checked it?
FWIW: I did a musl-cross-make cross compilation on an x86_64-linux-gnu host (https://github.com/llvm/llvm-project/issues/85504 help wanted to resolve a cmake issue) and `ninja check-llvm-transforms` passed.
